crates/dids/src/method/dht/bep44.rs

```diff
@@ -0,0 +1,233 @@
+};
+
+use byteorder::{BigEndian, ReadBytesExt, WriteBytesExt};
+use crypto::CryptoError;
+use keys::key::{KeyError, PublicKey};
+
+const MIN_MESSAGE_LEN: usize = 72;
+const MAX_MESSAGE_LEN: usize = 1072;
+
+/// Errors that can occur when working with Bep44 messages for did:dht.
+#[derive(thiserror::Error, Debug)]
+pub enum Bep44EncodingError {
+    #[error(transparent)]
+    SystemTimeError(#[from] SystemTimeError),
+    #[error("Could not coerce microseconds since epoch into u64")]
+    TimeError,
+    #[error(transparent)]
+    CryptoError(#[from] CryptoError),
+    #[error("Failure creating DID: {0}")]
+    BigEndianError(String),
+    #[error(
+        "Message must have size between {MIN_MESSAGE_LEN} and {MAX_MESSAGE_LEN} but got size {0}"
+    )]
+    SizeError(usize),
+    #[error(transparent)]
+    SignatureError(#[from] KeyError),
+}
+
+#[derive(Debug, PartialEq)]
+pub struct Bep44Message {
+    /// The sequence number of the message, used to ensure the latest version of
+    /// the data is retrieved and updated. It's a monotonically increasing number.
+    pub seq: u64,
+    /// The signature of the message, ensuring the authenticity and integrity
+    /// of the data. It's computed over the bencoded sequence number and value.
+    pub sig: Vec<u8>,
+    /// The actual data being stored or retrieved from the DHT network, typically
+    /// encoded in a format suitable for DNS packet representation of a DID Document.
+    pub v: Vec<u8>,
+}
+
+fn signable(seq: u64, message: &[u8]) -> Vec<u8> {
+    let mut signable = format!("3:seqi{}e1:v{}:", seq, message.len()).into_bytes();
+    signable.extend(message);
+    signable
+}
+
+fn encode_seq(seq: u64) -> Result<Vec<u8>, Bep44EncodingError> {
+    let mut seq_bytes = vec![];
+    seq_bytes.write_u64::<BigEndian>(seq).map_err(|_| {
+        Bep44EncodingError::BigEndianError("Failed to write big endian seq".to_string())
+    })?;
+    Ok(seq_bytes)
+}
+
+fn decode_seq(seq_bytes: &[u8]) -> Result<u64, Bep44EncodingError> {
+    let mut rdr = Cursor::new(seq_bytes);
+    let seq = rdr.read_u64::<BigEndian>().map_err(|_| {
+        Bep44EncodingError::BigEndianError("Failed to read big endian seq".to_string())
+    })?;
+    Ok(seq)
+}
+
+/// Represents a BEP44 message, which is used for storing and retrieving data
+/// in the Mainline DHT network.
+///
+/// A BEP44 message is used primarily in the context of the DID DHT method
+/// for publishing and resolving DID documents in the DHT network. This type
+/// encapsulates the data structure required for such operations in accordance
+/// with BEP44.
+///
+/// See [BEP44 Specification](https://www.bittorrent.org/beps/bep_0044.html)
+impl Bep44Message {
+    pub fn new<F>(message: &[u8], sign: F) -> Result<Self, Bep44EncodingError>
+    where
+        F: Fn(Vec<u8>) -> Result<Vec<u8>, KeyError>,
+    {
+        let seq: u64 = SystemTime::now()
+            .duration_since(UNIX_EPOCH)?
+            .as_micros()
+            .try_into()
```

Comment:
I'm not sure what this does but the spec has the seq in seconds https://did-dht.com/#create

Reply:
Ah that's a mistake because copied it blindly from web5-dart. Here's a PR to update it in dart https://github.com/TBD54566975/web5-dart/pull/87